src/coreclr/jit/morph.cpp

```diff
@@ -11266,7 +11266,7 @@ GenTree* Compiler::fgMorphSmpOp(GenTree* tree, MorphAddrContext* mac)
                     // castType is larger or the same as op1's type
                     // then we can discard the cast.
 
-                    if (varTypeIsSmall(castType) && (genTypeSize(castType) >= genTypeSize(effectiveOp1)))
+                    if (varTypeIsSmall(castType) && (genActualTypeSize(castType) == genActualTypeSize(effectiveOp1)))
```

Comment:
The comment above talks about this (well, it just states it as a fact I suppose) -- it is not correct to remove casts in case they would affect the upper bits after truncation.

Example: `ASG(IND<short>, CAST<ubyte>(x))`.